File: service/venue_service.go
package services

import (
	"cs-server/api/besttime"
	"cs-server/dao/redis"
	"cs-server/models/venue"
	"cs-server/util"
	"cs-server/config"
)

type VenueService struct {
	venueDao    *redis.RedisVenueDAO
	besttimeApi besttime.BestTimeAPI
}

// NewVenueService constructs a new VenueService with Redis dependency injection.
func NewVenueService(
	venueDao *redis.RedisVenueDAO,
	bestTimeApi besttime.BestTimeAPI) *VenueService {

	return &VenueService{
		venueDao:    venueDao,
		besttimeApi: bestTimeApi,
	}
}

func (vs *VenueService) GetVenuesNearby(lat, lon, radius float64) ([]venue.Venue, error) {
	return vs.venueDao.GetNearbyVenues(lat, lon, radius)
}

func (vs *VenueService) GetAllVenuesIds() ([]string, error) {
	return util.ReadVenuesIds(config.GetResourcePath(config.VENUES_IDS_RESOURCE))
}

func (vs *VenueService) GetVenue(venueId string) (*venue.Venue, error) {
	return vs.besttimeApi.GetVenue(venueId)
}

File: app/services/venue_service.py
"""Simple venue service wrapper over DAO."""
import logging
from typing import Optional

from app.api import BestTimeAPIClient
from app.dao import RedisVenueDAO
from app.models import Venue

logger = logging.getLogger(__name__)


class VenueService:
    """Service for venue queries (simple wrapper over DAO)."""

    def __init__(self, venue_dao: RedisVenueDAO, besttime_api: BestTimeAPIClient):
        """Initialize venue service.

        Args:
            venue_dao: Redis DAO for venue persistence
            besttime_api: BestTime API client
        """
        self.venue_dao = venue_dao
        self.besttime_api = besttime_api

    def get_venues_nearby(self, lat: float, lon: float, radius: float) -> list[Venue]:
        """Get venues within radius of a location.

        Args:
            lat: Latitude
            lon: Longitude
            radius: Radius in kilometers

        Returns:
            List of venues within radius
        """
        return self.venue_dao.get_nearby_venues(lat, lon, radius)
